flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -259,6 +255,62 @@ static fir::ExtendedValue arraySectionElementToExtendedValue(
   return fir::factory::componentToExtendedValue(builder, loc, element);
 }
 
+/// Convert the array_load, `load`, to an extended value. If `path` is not
+/// empty, then traverse through the components designated. The base value is
+/// `newBase`. This does not accept an array_load with a slice operand.
+static fir::ExtendedValue arrayLoadExtValue(fir::FirOpBuilder &builder,
+                                            mlir::Location loc,
+                                            fir::ArrayLoadOp load,
+                                            llvm::ArrayRef<mlir::Value> path,
+                                            mlir::Value newBase) {
+  // Recover the extended value from the load.
+  assert(!load.slice() && "slice is not allowed");
+  auto arrTy = load.getType();
+  auto idxTy = builder.getIndexType();
+  if (!path.empty()) {
+    auto ty = fir::applyPathToType(arrTy, path);
+    if (!ty)
+      fir::emitFatalError(loc, "path does not apply to type");
+    if (!ty.isa<fir::SequenceType>()) {
+      if (auto charTy = ty.dyn_cast<fir::CharacterType>()) {
+        // ???: Is this in CharacterExprHelper?
+        auto len = charTy.hasConstantLen()
+                       ? builder.createIntegerConstant(
+                             loc, idxTy, ty.cast<fir::CharacterType>().getLen())
+                       : load.typeparams()[0];
+        return fir::CharBoxValue{newBase, len};
+      }
+      return newBase;
+    }
+    arrTy = ty.cast<fir::SequenceType>();
+  }
+  // Recycle componentToExtendedValue if it looks plausible.
+  if (!fir::hasDynamicSize(arrTy))
+    return fir::factory::componentToExtendedValue(builder, loc, newBase);
+
```

Comment:
Is the code following this reachable in F95 ?
If my understanding of the function is correct, it seems wrong to me that it is using the shape of the array_load as being the shape of the element from the array_fetch/array_update. In this case, it might be better to just leave a single TODO below this.

Reply:
The array_update case returns an array value of the same shape as the array value being updated. This code is reached since we want to support assumed shape, etc.